### PR TITLE
Blackjack

### PR DESCRIPTION
**Pull request by ak2253**

### CODE REVIEW

File: server/routes/blackjack.py [hunk 1]
```diff
@@ -36,9 +65,29 @@ def play_blackjack():
...
+            .scalar()
+        )
+        if bet_amount > total_tickets:
+            return {
+                "success": False,
+                "message": "You are wagering more tickets than you currently have"
+            }
+        return {"success": True}
+    except json.decoder.JSONDecodeError:
+        return {"error": "Malformed request"}, 400
```
Comment (dvkm) [inline server/routes/blackjack.py:85]: Add lines between methods

File: server/routes/blackjack.py [hunk 2]
```diff
@@ -74,3 +123,112 @@ def bet_blackjack():
...
+
+@blackjack_bp.route("/api/blackjack/stand", methods=["GET"])
+def stand_blackjack():
+    query = Blackjack.query.filter_by(user_id=session["user_id"]).first()
+    deck = json.loads(query.deck)
+    player_hand = json.loads(query.player_hand)
+    dealer_hand = json.loads(query.dealer_hand)
+    while blackjack_total(dealer_hand) <= 17:
+        dealer_hand.append(draw_card(deck))
+    if len(deck)<200:
```
Comment (dvkm) [inline server/routes/blackjack.py]: `if len(deck) < 200:`

File: server/routes/blackjack.py [hunk 3]
```diff
@@ -74,3 +123,112 @@ def bet_blackjack():
...
+
+    client_dealer = translate_hand(dealer_hand)
+    client_player = translate_hand(player_hand)
+    dealer_total = blackjack_total(dealer_hand)
+    player_total = blackjack_total(player_hand)
+    if dealer_total > 21 or dealer_total < player_total:
+        last_transaction = db.session.query(Transaction)\
+            .filter_by(user_id=session["user_id"])\
+            .order_by(Transaction.id.desc()).first()
+        new_amount = last_transaction.ticket_amount * 1.5
```
Comment (dvkm) [inline server/routes/blackjack.py]: This is exploitable. A player could bet a small amount and if he knows he's about to win, he could bet a lot of tokens on coinflip. He can continue until he loses. Then, he'll get 1.5 times the last transaction amount.
For example, I bet 500 tickets on blackjack, when I get 21 and I'm likely to win, I bet 1,000,000 on coinflip. If I win, I continue until I lose. When blackjack is over, I'll get 1,500,000 tickets after betting just 500 tickets.

Also, why is the multiplier 1.5? Shouldn't it be 2? You earn the amount of money you bet (1) and the money you bet (1) = 1+1 = 2.

Comment (dvkm) [reply server/routes/blackjack.py]: Looked this up. Player blackjack is 1.5 but player vs house is 2.

Comment (dvkm) [reply server/routes/blackjack.py]: 2 makes more sense here since this is a dealer blackjack.

**Review by dvkm — CHANGES_REQUESTED** [review]

File: server/routes/blackjack.py (hunk 1 above)
Comment (ak2253) [reply server/routes/blackjack.py:85]: By methods do you mean line 83/84? or the return content?

Comment (dvkm) [reply server/routes/blackjack.py:85]: line break after line 85. New methods starts line 86.

File: server/routes/blackjack.py (hunk 3 above)
Comment (ak2253) [reply server/routes/blackjack.py]: I fixed it so that it also searches only for blackjack as activity in case someone tries that. The new gain is *2.

Comment (ak2253) [reply server/routes/blackjack.py]: Please check if that is good or there is another hole.

File: src/Blackjack.tsx [hunk 4]
```diff
@@ -1,27 +1,36 @@
...
+    <Typography variant="h6">Rules:</Typography>
+    <Box style={{ textAlign: 'left', paddingLeft: '40px' }}>
+      <p>1. Goal of blackjack is to beat the dealer&apos;s hand without going over 21</p>
+      <p>2. Jack, Queen and King are worth 10 points</p>
+      <p>3. Aces are worth 1 or 11 points. Which ever makes the hand better</p>
+      <p>4. Value cards are worth its repective value</p>
+      <p>7. &apos;Hit&apos; is to ask for another card.</p>
+      <p>8. &apos;Stand&apos; is to keep your total and end asking for cards</p>
+      <p>11. Dealer must hit until total is above 17</p>
+      <p>12. Reward is 1.5 worth of your total bet</p>
```
Comment (dvkm) [inline src/Blackjack.tsx]: 1.5 > 2

**Review by dvkm — APPROVED** [review]

**Review by speshori — APPROVED** [review]

File: src/Blackjack.tsx (hunk 4 above)
Comment (ak2253) [reply src/Blackjack.tsx]: Made change

**Pull request merged**